auth/auth.go

```diff
@@ -0,0 +1,157 @@
+package auth
+
+import (
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/pkg/errors"
+)
+
+// Interface for authenticating and retrieving users.
+type Interface interface {
+	Authenticate(username, password string) (User, error)
+	User(username string) (User, error)
+}
+
+type Privilege int
+
+const (
+	NoPrivileges Privilege = 1 << iota
+
+	ReadPrivilege
+	WritePrivilege
+	DeletePrivilege
+
+	AllPrivileges
+)
+
+func (p Privilege) String() string {
+	switch p {
+	case NoPrivileges:
+		return "NO_PRIVILEGES"
+	case ReadPrivilege:
+		return "read"
+	case WritePrivilege:
+		return "write"
+	case DeletePrivilege:
+		return "delete"
+	case AllPrivileges:
+		return "ALL_PRIVILEGES"
+	default:
+		return "UNKNOWN_PRIVILEGE"
+	}
+}
+
+type Action struct {
+	Resource string
+	Method   string
+}
+
+func (a Action) RequiredPrivilege() (Privilege, error) {
+	switch m := strings.ToUpper(a.Method); m {
+	case "HEAD", "OPTIONS":
+		return NoPrivileges, nil
+	case "GET":
+		return ReadPrivilege, nil
+	case "POST", "PATCH":
+		return WritePrivilege, nil
+	case "DELETE":
+		return DeletePrivilege, nil
+	default:
+		return AllPrivileges, fmt.Errorf("unknown method %q", m)
+	}
+}
```

Comment:
Wondering if authorizing the HTTP request is the right way to go or if it should be somewhere below the transport layer. Thoughts?